### PR TITLE
Calc temperature after latent heat release

### LINKED ISSUE #1738: Method to iteratively calculate saturated adiabatic adjustment from release of latent heat of condensation
As a science lead I want a method to return the temperature after latent heat release so that I can use this in future complex convection calculations with the end point of generating hail accumulation for use with MOGREPS-G data.

This will be applied to data after the dry adiabatic adjustment implemented in #1730. The inputs are temperature and specific humidity, both of which will be modified and returned.

image

Acceptance criteria:
 * Fast iterative solver implemented for this method
 * Works for saturated, near-saturated and sub-saturated conditions

### PR DESCRIPTION
**Pull request by MoseleyS**
Closes #1738 

Adds method to adjust temperature and humidity for latent heat release.

- Moves `WetBulbTemperature` and `WetBulbTemperatureIntegral` plugins into a new `wet_bulb_temperature.py` module as `psychrometric_calculations` is getting too big.
- Moves `_calculate_latent_heat` method from `WetBulbTemperature` plugin to `psychrometric_calculations` so that it can be used in latent heat release calculations.
- Adds extra tests for dry adiabatic methods to cover a higher altitude case.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

**Review by bayliffe — CHANGES_REQUESTED** [review]
A few requests. I might play with this some more when it comes back to me, but I think it is working as expected. Extended documentation around any new methods would be great; we could really improve the docs around psychrometric calculations more generally, but that's future work.

File: improver/psychrometric_calculations/psychrometric_calculations.py [hunk 1]
```diff
@@ -177,373 +175,108 @@ def dry_adiabatic_pressure(
...
-        Met Office UM Documentation Paper 080, UM Version 10.8,
-        last updated 2014-12-05.
+    Returns:
+        Array of specific humidity values (kg kg-1) representing saturated air
     """
+    epsilon = 0.622
+    svp = calculate_svp_in_air(temperature, pressure)
+    return (epsilon * svp / (pressure - (1.0 - epsilon * svp))).astype(
+        temperature.dtype
+    )
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver/psychrometric_calculations/psychrometric_calculations.py:195]: I hadn't spotted that they were so similar. I've refactored this so there is now only one method - this slightly altered the test results. I've also repeated the test values in the discarded method and found that one of them wasn't right - the requirement to be accurate to 7 decimal places wasn't appropriate to a value < 1e-7.

File: improver/psychrometric_calculations/psychrometric_calculations.py [hunk 2]
```diff
@@ -177,373 +175,108 @@ def dry_adiabatic_pressure(
...
-            cubelist.append(
-                self.create_wet_bulb_temperature_cube(
-                    t_slice.copy(), rh_slice.copy(), p_slice.copy()
-                )
-            )
-        wet_bulb_temperature = cubelist.merge_cube()
+def _latent_heat_release(q1: ndarray, q2: ndarray, temperature: ndarray) -> ndarray:
+    """Returns the latent heat released (K) when condensing water vapour from specific humidity
+    value q1 to q2, both in kg kg-1 when the temperature is approximately t.
+    Returns negative values when initial condition is subsaturated.
```
[the remark this answers is not in the document]
Comment (bayliffe) [reply improver/psychrometric_calculations/psychrometric_calculations.py]: That's useful, thanks.

**Comment by bayliffe** [comment]
@MoseleyS can you fix the broken unit tests please, and generate a ticket about extending the psychrometric calculations documentation for future work.

File: improver_tests/psychrometric_calculations/test_psychrometric_calculations.py [hunk 3]
```diff
@@ -64,3 +75,66 @@ def test_round_trip(shape, method, t1, p1, n2, expected):
...
+    """Test the saturated_humidity method"""
+    result = saturated_humidity(
+        np.full(shape, t, dtype=np.float32), np.full(shape, p, dtype=np.float32)
+    )
+    assert np.isclose(result, expected).all()
+    assert result.shape == shape
+    assert result.dtype == np.float32
+
+
+# Stephen checked these values on a Tephigram.
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver_tests/psychrometric_calculations/test_psychrometric_calculations.py:106]: I used this image: https://en.wikipedia.org/wiki/Tephigram#/media/File:Tephigram.png

File: improver/psychrometric_calculations/psychrometric_calculations.py [hunk 4]
```diff
@@ -177,373 +175,108 @@ def dry_adiabatic_pressure(
...
+        q1:
+            Specific humidity before latent heat release (kg kg-1)
+        q2:
+            Specific humidity after latent heat release (kg kg-1)
 
-        return wet_bulb_temperature
+    Returns:
+        Temperature adjustment to apply to account for latent heat release (K).
+    """
+    return (_calculate_latent_heat(temperature) / consts.CP_WATER_VAPOUR) * (q1 - q2)
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver/psychrometric_calculations/psychrometric_calculations.py:240]: I've opened a ticket for this, and made a start on it too. https://github.com/metoppv/improver/issues/1746

**Review by bayliffe — DISMISSED** [review]

File: improver_tests/psychrometric_calculations/wet_bulb_temperature/test_WetBulbTemperature.py [hunk 5]
```diff
@@ -55,21 +53,6 @@ def setUp(self):
...
-
-    def test_calculate_mixing_ratio(self):
-        """Test mixing ratio calculation"""
-        pressure = np.array([1.0e5, 9.9e4, 9.8e4], dtype=np.float32)
-        expected = [6.06744631e-08, 1.31079322e-03, 1.77063149e-01]
-        result = WetBulbTemperature()._calculate_mixing_ratio(
-            self.temperature, pressure
-        )
-        self.assertArrayAlmostEqual(result, expected, decimal=7)
-
```
Comment (Kat-90) [inline improver_tests/psychrometric_calculations/wet_bulb_temperature/test_WetBulbTemperature.py:72]: No problem with this section of code, but I was checking the rest of it had everything removed that needed to be (couldn't comment on that line). On line 94, the mixing ratio cube is set for the new Test_WetBulbTemperature class and I don't think it's ever used in that class.

Comment (MoseleyS) [reply improver_tests/psychrometric_calculations/wet_bulb_temperature/test_WetBulbTemperature.py:72]: You're right. This is never used. I've removed it.

**Review by Kat-90 — APPROVED** [review]
Checked the code/documentation and unit tests and ran all tests.

Happy to approve.

**Pull request merged**